### PR TITLE
Add Lyrics Editor for Admin users only

### PR DESCRIPTION
**Pull request by scampower3**
Adds a lyrics editor dialog so that admin users can manually add, remove or search for lyrics through the web ui. Code is based on the subtitle editor.

Search has been tested using the [LrcLib plugin](https://github.com/crobibero/jellyfin-plugin-lrclib)

**Changes**

image
image
image
image

**Issues**

### CODE REVIEW

File: src/components/lyricseditor/lyricseditor.js [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+import template from './lyricseditor.template.html';
+import templatePreview from './lyricspreview.template.html';
+import { deleteLyrics } from '../../scripts/deleteHelper';
+
+let currentItem;
+let hasChanges;
+
+function downloadRemoteLyrics(context, id) {
+    const api = toApi(ServerConnections.getApiClient(currentItem.ServerId));
+    const lyricsApi = new LyricsApi(api.configuration, undefined, api.axiosInstance);
```
Comment (thornbill) [inline src/components/lyricseditor/lyricseditor.js:30]: You should use the `getLyricsApi(api)` utility.

Comment (scampower3) [reply src/components/lyricseditor/lyricseditor.js:30]: I think that wasn't present or I couldn't find it when I initally coded this 🙃
Will update to use it.

Comment (thornbill) [reply src/components/lyricseditor/lyricseditor.js]: It probably hadn't been added yet 😅

**Comment by thornbill** [comment]
@scampower3 do you think there is a way we could limit some of the code duplication between this and the subtitle editor?

**Comment by scampower3** [comment]
@thornbill I don't think so. The highlighted duplicates by sonarcloud are mostly the event handlers and its associated logic.

File: src/components/lyricseditor/lyricseditor.template.html [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            </div>
+            <button is="emby-button" type="submit" class="raised btnSubmit block button-submit">${Search}</button>
+        </form>
+        <br />
+        <div class="lyricsResults"></div>
+        <div class="noSearchResults hide">
+            ${NoLyricsSearchResultsFound}
+        </div>
+    </div>
+</div>
```
Comment (thornbill) [inline src/components/lyricseditor/lyricseditor.template.html:27]: Files should have trailing newlines.

File: src/components/lyricseditor/lyricspreview.template.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<div class="formDialogHeader">
+    <button is="paper-icon-button-light" class="btnCancel autoSize" tabindex="-1" title="${ButtonBack}"><span class="material-icons arrow_back" aria-hidden="true"></span></button>
+    <h3 class="formDialogHeaderTitle">${HeaderPreviewLyrics}</h3>
+
+</div>
+<div class="formDialogContent smoothScrollY">
+    <div class="dialogContentInner dialog-content-centered">
+        <div class="lyricsPreview" style="margin-bottom:2em;"></div>
+    </div>
+</div>
```
Comment (thornbill) [inline src/components/lyricseditor/lyricspreview.template.html:10]: Files should have trailing newlines.

File: src/components/lyricsuploader/lyricsuploader.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+export function show(options) {
+    return new Promise(function (resolve) {
+        hasChanges = false;
+        showEditor(options, resolve);
+    });
+}
+
+export default {
+    show: show
```
Comment (thornbill) [inline src/components/lyricsuploader/lyricsuploader.js:170]:
```suggestion
    show
```

**Review by thornbill — CHANGES_REQUESTED** [review]

File: src/components/lyricseditor/lyricseditor.js [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+
+function onDeleteLyrics(e) {
+    deleteLyrics(currentItem).then(() => {
+        hasChanges = true;
+        const context = dom.parentWithClass(e.target, 'formDialogContent');
+        const apiClient = ServerConnections.getApiClient(currentItem.ServerId);
+        reload(context, apiClient, currentItem.Id);
+    }).catch(() => {
+        console.warn('Failed to delete lyrics for', currentItem.Name);
+    });
```
Comment (thornbill) [inline src/components/lyricseditor/lyricseditor.js:295]: I should have been more clear here. The `deleteLyrics` modal already handles if an error occurs while deleting, but our modal code uses kind of a bad pattern where closing a modal rejects, so really we just need to ignore it.

```suggestion
    }).catch(() => {
        // delete dialog closed
    });
```

Comment (scampower3) [reply src/components/lyricseditor/lyricseditor.js]: Yeah, that's what I assumed. But doing nothing in a catch statement doesn't really feel right to me, so I added some console logs there.😆

File: src/components/lyricseditor/lyricseditor.js [hunk 6]
```diff
@@ -0,0 +1,393 @@
+import escapeHtml from 'escape-html';
+
+import { getLyricsApi } from '@jellyfin/sdk/lib/utils/api/lyrics-api';
+import { toApi } from 'utils/jellyfin-apiclient/compat';
+import dialogHelper from '../dialogHelper/dialogHelper';
+import layoutManager from '../layoutManager';
+import globalize from '../../scripts/globalize';
```
Comment (thornbill) [inline src/components/lyricseditor/lyricseditor.js:7]:
```suggestion
import globalize from 'lib/globalize';
```

File: src/components/lyricsuploader/lyricsuploader.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+import { getLyricsApi } from '@jellyfin/sdk/lib/utils/api/lyrics-api';
+import { toApi } from 'utils/jellyfin-apiclient/compat';
+import dialogHelper from '../../components/dialogHelper/dialogHelper';
+import ServerConnections from '../ServerConnections';
+import dom from '../../scripts/dom';
+import loading from '../../components/loading/loading';
+import scrollHelper from '../../scripts/scrollHelper';
+import layoutManager from '../layoutManager';
+import globalize from '../../scripts/globalize';
```
Comment (thornbill) [inline src/components/lyricsuploader/lyricsuploader.js:11]:
```suggestion
import globalize from 'lib/globalize';
```

**Review by thornbill — APPROVED** [review]

**Pull request merged**